core/src/main/java/org/apache/calcite/plan/volcano/VolcanoPlanner.java

```diff
@@ -1117,22 +1117,27 @@ public void setImportance(RelNode rel, double importance) {
    * @see #normalizePlan(String)
    */
   public void dump(PrintWriter pw) {
-    pw.println("Root: " + root);
-    pw.println("Original rel:");
+    try {
+      pw.println("Root: " + root);
+      pw.println("Original rel:");
 
-    if (originalRoot != null) {
-      originalRoot.explain(
-          new RelWriterImpl(pw, SqlExplainLevel.ALL_ATTRIBUTES, false));
-    }
-    if (CalciteSystemProperty.DUMP_SETS.value()) {
-      pw.println();
-      pw.println("Sets:");
-      dumpSets(pw);
-    }
-    if (CalciteSystemProperty.DUMP_GRAPHVIZ.value()) {
-      pw.println();
-      pw.println("Graphviz:");
-      dumpGraphviz(pw);
+      if (originalRoot != null) {
+        originalRoot.explain(
+            new RelWriterImpl(pw, SqlExplainLevel.ALL_ATTRIBUTES, false));
+      }
+
+      if (CalciteSystemProperty.DUMP_SETS.value()) {
+        pw.println();
+        pw.println("Sets:");
+        dumpSets(pw);
+      }
+      if (CalciteSystemProperty.DUMP_GRAPHVIZ.value()) {
+        pw.println();
+        pw.println("Graphviz:");
+        dumpGraphviz(pw);
+      }
+    } catch (Exception ex) {
+      throw new AssertionError("Dumps the internal state ", ex);
```

Comment:
If we change this way, we won't see CannotPlanException

Reply:
you shouldn't throw a exception or error here.

Reply:
+1